### DISCUSSION TITLE
Legend State chrome local storage

### DISCUSSION THREAD

**Discussion by JohnBra** [LegendApp/legend-state#393] [Q&A]
Hi @jmeistrich, I love what you are doing here with legend-state! Great library 🚀

**TL;DR**: I was wondering if it is possible to use legend state to persist to the chrome extension local storage as opposed to the web local storage. We might need to write an `async` version of the [`local-storage.ts`](https://github.com/LegendApp/legend-state/blob/main/src/persist-plugins/local-storage.ts) Persistence Plugin accessing `chrome.storage.<'local'/'sync'>` rather than `window.localStorage` and then calling the corresponding `set/get` item functions.

I'm evaluating whether I can integrate legend-state with a Chrome extension I'm building to leverage the offline first behavior in combination with Supabase.

Before opening this discussion I tried a few things including:
- Look for similar issues with chrome extensions (none found)
- Use local storage which worked fine, but persisted to `window.localStorage` which isn't accessible from the content script
- Use indexeddb which I couldn't get to work. Transaction errors, i.e. `[legend-state] Error loading IndexedDB InvalidAccessError: Failed to execute 'transaction' on 'IDBDatabase': The storeNames parameter was empty.` _TBH, this could be a skill issue on my end, but I couldn't really make sense of the docs._ The [IndexedDB example](https://legendapp.com/open-source/state/v3/sync/persist-sync/#indexeddb-react) doesn't compile (see screenshots below)
<img width="400" alt="Screenshot 2024-11-11 at 4 04 57 PM" src="https://github.com/user-attachments/assets/3fb52797-eccd-428e-829a-6902e1063635">
<img width="1113" alt="Screenshot 2024-11-11 at 4 05 05 PM" src="https://github.com/user-attachments/assets/047bfdfd-23f1-4201-8291-ce8f76241886">

Which I can fix by flattening the options/removing the indexedDB prop, but will then get DB access errors. Maybe @jmeistrich has an idea what I'm doing wrong.

Optimally, I would like to use `chrome.storage.local` or `chrome.storage.sync` anyway. While the API is mostly the same, it differs a little from the `window.localStorage` in that:
- it's asynchronous
- `get/set` methods have different signatures
- it can save primitive/json type data directly without the need for serialization

What would be the easiest way to make this work?
I'm happy to contribute a persist plugin for the chrome storage according to [this interface](https://github.com/LegendApp/legend-state/blob/main/src/sync/syncTypes.ts#L133) if need be.
